### PR TITLE
[Workspace] Copy selected/all saved objects

### PR DESCRIPTION
**Pull request by yubonluo**
### Description

This PR follows up https://github.com/opensearch-project/OpenSearch-Dashboards/pull/6288 which can be used to duplicate saved objects among workspaces.
Only UI change in this PR.

### Issues Resolved

https://github.com/opensearch-project/OpenSearch-Dashboards/issues/6388

## Screenshot

https://github.com/opensearch-project/OpenSearch-Dashboards/assets/71875999/89fe15e1-c876-4f7c-b64a-77a5c550002c

There are three types of duplicate buttons on the saved objects management page.
image

If the user clicks the duplicate button, a modal window will pop up as follows:
image

## Changelog
- feat: [Workspace] Duplicate selected/all saved objects

## Testing the changes
1. Clone the latest code and run yarn osd bootstrap
2. Modify config/opensearch_dashboards.yml, add workspace.enabled: true
3. Run yarn start --no-base-path
4. Create a workspace and create some saved objects like index patterns, visualizations, etc.
5. Call the copy api following the above usage.

### Check List

- [x] All tests pass
  - [x] `yarn test:jest`
  - [x] `yarn test:jest_integration`
- [x] New functionality includes testing.
- [x] New functionality has been documented.
- [x] Update [CHANGELOG.md](./../CHANGELOG.md)
- [x] Commits are signed per the DCO using --signoff

### CODE REVIEW

**Comment by kavilla** [comment]
Hey @yubonluo, thanks for opening. I'm not able to see the videos.

Do you have some screen shots? Also just confirming with @opensearch-project/opensearch-ux the expected workflow.

**Comment by lauralexis** [comment]
> Hey @yubonluo, thanks for opening. I'm not able to see the videos.
> 
> Do you have some screen shots? Also just confirming with @opensearch-project/opensearch-ux the expected workflow.

@yubonluo I’d like to make some adjustments to the workflow you shared earlier. I’m putting my recommendation together.

**Comment by yubonluo** [comment]
> Hey @yubonluo, thanks for opening. I'm not able to see the videos.
> 
> Do you have some screen shots? Also just confirming with @opensearch-project/opensearch-ux the expected workflow.

The screen shots has been added.

File: src/plugins/workspace/server/saved_objects/workspace_saved_objects_client_wrapper.ts [hunk 1]
```diff
@@ -412,7 +413,7 @@ export class WorkspaceSavedObjectsClientWrapper {
     };
 
     const findWithWorkspacePermissionControl = async <T = unknown>(
-      options: SavedObjectsFindOptions
+      options: SavedObjectsFindOptions & Pick<WorkspaceFindOptions, 'permissionModes'>
```
Comment (wanglam) [inline src/plugins/workspace/server/saved_objects/workspace_saved_objects_client_wrapper.ts:416]: Why we need to add `permissionModes` to options? I think `permissionModes` is already exists in `options.ACLSearchParams`.

File: src/plugins/workspace/public/workspace_client.ts [hunk 2]
```diff
@@ -118,7 +120,20 @@ export class WorkspaceClient {
     });
 
     if (result?.success) {
-      this.workspaces.workspaceList$.next(result.result.workspaces);
+      const resultWithWritePermission = await this.list({
+        perPage: 999,
+        permissionModes: [WorkspacePermissionMode.LibraryWrite],
```
Comment (wanglam) [inline src/plugins/workspace/public/workspace_client.ts:125]: To support `permissionModes` parameters, I think we need to modify `src/plugins/workspace/server/workspace_client.ts` and pass `ACLSearchParams:{permissionModes}` to saved object find method.

File: src/plugins/saved_objects_management/public/lib/duplicate_saved_objects.ts [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import { HttpStart } from 'src/core/public';
+
+export async function duplicateSavedObjects(
+  http: HttpStart,
+  objects: any[],
+  includeReferencesDeep: boolean = true,
```
[the remark this answers is not in the document]
Comment (yubonluo) [reply src/plugins/saved_objects_management/public/lib/duplicate_saved_objects.ts:11]: Sure, thanks your advice, I will move  `includeReferencesDeep: boolean = true` to the last parameter.

File: src/plugins/saved_objects_management/public/lib/duplicate_saved_objects.test.ts [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+        "results": Array [
+          Object {
+            "type": "return",
+            "value": undefined,
+          },
+        ],
+      }
+    `);
+
+    await duplicateSavedObjects(httpClient, objects, undefined, targetWorkspace);
```
[the remark this answers is not in the document]
Comment (yubonluo) [reply src/plugins/saved_objects_management/public/lib/duplicate_saved_objects.test.ts:38]: Sure, I will separate two tests cases.

File: src/plugins/saved_objects_management/public/management_section/objects_table/components/duplicate_modal.tsx [hunk 5]
```diff
@@ -0,0 +1,232 @@
...
+    targetWorkspace: string,
+    targetWorkspaceName: string
+  ) => Promise<void>;
+  http: HttpSetup;
+  workspaces: WorkspacesStart;
+  notifications: NotificationsStart;
+  selectedSavedObjects: DuplicateObject[];
+}
+
+interface Props extends ShowDuplicateModalProps {
```
[the remark this answers is not in the document]
Comment (yubonluo) [reply src/plugins/saved_objects_management/public/management_section/objects_table/components/duplicate_modal.tsx:40]: Sure, I will modify the logic.

File: src/plugins/saved_objects_management/public/management_section/objects_table/saved_objects_table.tsx [hunk 6]
```diff
@@ -235,6 +274,9 @@ export class SavedObjectsTable extends Component<SavedObjectsTableProps, SavedOb
     this._isMounted = false;
     this.debouncedFetchObjects.cancel();
     this.unSubscribeWorkspace();
+    this.currentWorkspaceIdSubscription?.unsubscribe();
+    this.workspacesSubscription?.unsubscribe();
```
[the remark this answers is not in the document]
Comment (yubonluo) [reply src/plugins/saved_objects_management/public/management_section/objects_table/saved_objects_table.tsx:278]: Sure, I will remove this code.

File: src/plugins/saved_objects_management/public/management_section/objects_table/saved_objects_table.tsx [hunk 7]
```diff
@@ -235,6 +274,9 @@ export class SavedObjectsTable extends Component<SavedObjectsTableProps, SavedOb
...
+    this.workspacesEnabledSubscription?.unsubscribe();
```
[the remark this answers is not in the document]
Comment (yubonluo) [reply src/plugins/saved_objects_management/public/management_section/objects_table/saved_objects_table.tsx:279]: Sure, I will remove the useless code.

File: src/plugins/saved_objects_management/public/management_section/objects_table/components/table.tsx [hunk 8]
```diff
@@ -62,14 +63,15 @@ export interface TableProps {
...
   columnRegistry: SavedObjectsManagementColumnServiceStart;
-  namespaceRegistry: SavedObjectsManagementNamespaceServiceStart;
   selectedSavedObjects: SavedObjectWithMetadata[];
   selectionConfig: {
     onSelectionChange: (selection: SavedObjectWithMetadata[]) => void;
   };
   filters: any[];
   canDelete: boolean;
   onDelete: () => void;
+  onDuplicateSelected: () => void;
```
[the remark this answers is not in the document]
Comment (yubonluo) [reply src/plugins/saved_objects_management/public/management_section/objects_table/components/table.tsx:73]: Sure, I will rename it `onDuplicateSelected` to `onDuplicate `.

File: src/plugins/saved_objects_management/public/management_section/objects_table/components/table.tsx [hunk 9]
```diff
@@ -62,14 +63,15 @@ export interface TableProps {
   basePath: IBasePath;
   actionRegistry: SavedObjectsManagementActionServiceStart;
   columnRegistry: SavedObjectsManagementColumnServiceStart;
-  namespaceRegistry: SavedObjectsManagementNamespaceServiceStart;
```
[the remark this answers is not in the document]
Comment (yubonluo) [reply src/plugins/saved_objects_management/public/management_section/objects_table/components/table.tsx:65]: I found that `namespaceRegistry ` was not used and was highlighted in red, so I deleted it. I will recover the code.

File: src/plugins/saved_objects_management/public/management_section/objects_table/saved_objects_table.tsx [hunk 10]
```diff
@@ -152,6 +155,7 @@ export class SavedObjectsTable extends Component<SavedObjectsTableProps, SavedOb
   private _isMounted = false;
   private currentWorkspaceIdSubscription?: Subscription;
   private workspacesSubscription?: Subscription;
+  private workspacesEnabledSubscription?: Subscription;
```
[the remark this answers is not in the document]
Comment (yubonluo) [reply src/plugins/saved_objects_management/public/management_section/objects_table/saved_objects_table.tsx:158]: Sure, I will remove it.

File: src/plugins/saved_objects_management/public/management_section/objects_table/saved_objects_table.tsx [hunk 11]
```diff
@@ -658,6 +674,134 @@ export class SavedObjectsTable extends Component<SavedObjectsTableProps, SavedOb
...
+      result = await duplicateSavedObjects(
+        http,
+        objectsToDuplicate,
+        includeReferencesDeep,
+        targetWorkspace
+      );
+      const successCount = result.success
+        ? savedObjects.length
+        : savedObjects.length - result.errors.length;
+      const sOMessage =
```
[the remark this answers is not in the document]
Comment (yubonluo) [reply src/plugins/saved_objects_management/public/management_section/objects_table/saved_objects_table.tsx:729]: Sure, I will rename `sOMessage` to `titleOrCountMessage`.

File: src/plugins/saved_objects_management/public/management_section/objects_table/saved_objects_table.tsx [hunk 12]
```diff
@@ -658,6 +674,134 @@ export class SavedObjectsTable extends Component<SavedObjectsTableProps, SavedOb
...
+                errorIdMessages,
+            }
+          ),
+        });
+      }
+    } catch (e) {
+      notifications.toasts.addDanger({
+        title: i18n.translate('savedObjectsManagement.objectsTable.duplicate.dangerNotification', {
+          defaultMessage:
+            'Error - Unable to duplicate ' + savedObjects.length.toString() + ' saved object(s)',
```
[the remark this answers is not in the document]
Comment (yubonluo) [reply src/plugins/saved_objects_management/public/management_section/objects_table/saved_objects_table.tsx:778]: Sure, I will add a simple variable.

File: src/plugins/saved_objects_management/public/management_section/objects_table/saved_objects_table.tsx [hunk 13]
```diff
@@ -658,6 +674,134 @@ export class SavedObjectsTable extends Component<SavedObjectsTableProps, SavedOb
...
+        successCount === 1
+          ? savedObjects[0].meta.title
+          : successCount.toString() + ' saved objects';
+      const refMessage = includeReferencesDeep ? ' and the related objects' : '';
+      if (result.success) {
+        notifications.toasts.addSuccess({
+          title: i18n.translate(
+            'savedObjectsManagement.objectsTable.duplicate.successNotification',
+            {
+              defaultMessage:
```
[the remark this answers is not in the document]
Comment (yubonluo) [reply src/plugins/saved_objects_management/public/management_section/objects_table/saved_objects_table.tsx:739]: Sure, I will use the select synax.

File: src/plugins/workspace/public/workspace_client.ts (hunk 2 above)
Comment (yubonluo) [reply src/plugins/workspace/public/workspace_client.ts:125]: Sure, I will add `ACLSearchParams: { permissionModes: options.permissionModes }` to find object in the `src/plugins/workspace/server/workspace_client.ts`.

File: src/plugins/workspace/public/workspace_client.ts [hunk 14]
```diff
@@ -118,7 +120,20 @@ export class WorkspaceClient {
...
+      });
+      if (resultWithWritePermission?.success) {
+        const workspaceIdsWithWritePermission = resultWithWritePermission.result.workspaces.map(
+          (workspace: WorkspaceAttribute) => workspace.id
+        );
+        const workspaces = result.result.workspaces.map((workspace: WorkspaceAttribute) => ({
+          ...workspace,
+          libraryReadonly: !workspaceIdsWithWritePermission.includes(workspace.id),
```
[the remark this answers is not in the document]
Comment (yubonluo) [reply src/plugins/workspace/public/workspace_client.ts:133]: I will remove `libraryReadonly `.

File: src/plugins/workspace/server/saved_objects/workspace_saved_objects_client_wrapper.ts (hunk 1 above)
Comment (yubonluo) [reply src/plugins/workspace/server/saved_objects/workspace_saved_objects_client_wrapper.ts:416]: I will remove the useless code.

**Comment by yubonluo** [comment]
CC @SuZhou-Joe @ruanyl could you please help me review this pr?

**Comment by BionIT** [comment]
Hi @yubonluo , are we targeting this PR for 2.15 or 2.16?

**Comment by yubonluo** [comment]
> Hi @yubonluo , are we targeting this PR for 2.15 or 2.16?

Hi @BionIT due to waiting for the final confirmation of the UI text, this PR will be released on 2.16.

**Review by ruanyl — APPROVED** [review]
Overall looks good to me, left some comments, should not be blocker

**Review by ruanyl — APPROVED** [review]

**Pull request merged**